File: scripts/camera-controls.ts
export type CameraControl = {
  name: string;
  value: string;
};

export const defaultCameraControls: CameraControl[] = [
  { name: "brightness", value: "-5" },
  { name: "contrast", value: "1" },
  { name: "saturation", value: "64" },
  { name: "white_balance_automatic", value: "0" },
  { name: "white_balance_temperature", value: "4600" },
  { name: "gamma", value: "100" },
  { name: "gain", value: "32" },
  { name: "power_line_frequency", value: "1" },
  { name: "sharpness", value: "1" },
  { name: "backlight_compensation", value: "0" },
  { name: "auto_exposure", value: "1" },
  { name: "exposure_time_absolute", value: "10" },
  { name: "focus_automatic_continuous", value: "0" },
  { name: "focus_absolute", value: "0" },
];

export function applyDefaultCameraControls(devices: readonly [string, string], cwd: string): void {
  for (const command of buildCameraControlCommands(devices)) {
    const result = Bun.spawnSync(command, {
      cwd,
      env: process.env,
      stdout: "inherit",
      stderr: "inherit",
    });
    if (result.exitCode !== 0) {
      throw new Error(`Failed to apply camera controls: ${displayCameraControlCommand(command)}`);
    }
  }
}

export function buildCameraControlCommands(devices: readonly [string, string]): string[][] {
  return devices.flatMap((device) =>
    defaultCameraControls.map((control) => ["v4l2-ctl", "-d", device, `--set-ctrl=${control.name}=${control.value}`]),
  );
}

export function displayCameraControlCommand(command: string[]): string {
  return command.map(shellWord).join(" ");
}

function shellWord(value: string): string {
  return /^[A-Za-z0-9_./:=,@+-]+$/.test(value) ? value : `'${value.replaceAll("'", "'\\''")}'`;
}

File: scripts/camera_controls.py
#!/usr/bin/env python3

from __future__ import annotations

import subprocess
from pathlib import Path
from typing import Sequence

from process_utils import display_command, process_env


DEFAULT_CAMERA_CONTROLS: tuple[tuple[str, str], ...] = (
    ("brightness", "-5"),
    ("contrast", "1"),
    ("saturation", "64"),
    ("white_balance_automatic", "0"),
    ("white_balance_temperature", "4600"),
    ("gamma", "100"),
    ("gain", "32"),
    ("power_line_frequency", "1"),
    ("sharpness", "1"),
    ("backlight_compensation", "0"),
    ("auto_exposure", "1"),
    ("exposure_time_absolute", "10"),
    ("focus_automatic_continuous", "0"),
    ("focus_absolute", "0"),
)


def build_camera_control_commands(devices: Sequence[str]) -> list[list[str]]:
    return [
        ["v4l2-ctl", "-d", device, f"--set-ctrl={name}={value}"]
        for device in devices
        for name, value in DEFAULT_CAMERA_CONTROLS
    ]


def display_camera_control_command(command: Sequence[str]) -> str:
    return display_command(command)


def apply_default_camera_controls(devices: Sequence[str], cwd: Path) -> None:
    for command in build_camera_control_commands(devices):
        result = subprocess.run(command, cwd=cwd, env=process_env(), check=False)
        if result.returncode != 0:
            raise RuntimeError(
                f"Failed to apply camera controls: {display_camera_control_command(command)}"
            )
